### PR TITLE
Corrected Docker and Make files. Added test automation scripts.

### PR DESCRIPTION
**Pull request by devang-gaur**
fabric8-hdd-bug-id: fabric8-hdd/supervisor#28
Signed-off-by: Devang Gaur <dgaur@redhat.com>

### CODE REVIEW

**Comment by mik-dass** [comment]
LGTM

File: Makefile
Comment (anmolbabu) [inline Makefile]: Undo this change I think you have this for the purpose of verification but have a look at https://github.com/fabric8-hdd/supervisor/pull/38 we are moving away from the private repos.

File: run_tests.sh
Comment (anmolbabu) [inline run_tests.sh:18]: Please get an ack from @mitchharpur or @miteshvp

File: src/api/routes.js
Comment (anmolbabu) [inline src/api/routes.js]: I was planning / to be a kind of possible routes descriptor.
Like it should list all possible routes supported by supervisor at any given point in time along with the route description like params, response and route description like an rsdl in java applications

File: package.json
Comment (anmolbabu) [inline package.json:22]: This was a generated value.
Please tell me what was the reason for this change?

File: cico_setup.sh
Comment (anmolbabu) [inline cico_setup.sh]: Please leave these changes to https://github.com/fabric8-hdd/supervisor/pull/38

File: cico_setup.sh
Comment (anmolbabu) [inline cico_setup.sh]: Please leave these changes to #38

File: cico_setup.sh
Comment (anmolbabu) [inline cico_setup.sh]: Please leave these changes to #38

File: cico_setup.sh
Comment (anmolbabu) [inline cico_setup.sh]: Unintentional change?

File: run_tests.sh
Comment (anmolbabu) [reply run_tests.sh:18]: Please refer the thread around https://github.com/fabric8-hdd/supervisor/pull/29#discussion_r155695795

Comment (anmolbabu) [reply run_tests.sh:18]: Our stress tests, n/w tests, etc... can probably be executed here

File: cico_setup.sh
Comment (devang-gaur) [reply cico_setup.sh]: I've shifted this echo part to the push image function so that we can verify if the push actually happened with 
if [ $? -eq 0 ] ...success else exit 1

Earlier, the success method used to be echoed regardless of whether a push happened successfully or not.

File: package.json
Comment (devang-gaur) [reply package.json:22]: I was recieving npm warn console logs while running the scripts.

I referred to this document : https://gist.github.com/robertkowalski/7620849

and changed it to the string with value 133.

File: src/api/routes.js
Comment (devang-gaur) [reply src/api/routes.js]: Sure, we can change the response message for that.
Right now , when I do a GET / . Nothing appeared on the browser. We probably need to use nginx for route descripter page at this route.

File: cico_setup.sh
Comment (devang-gaur) [reply cico_setup.sh]: Hey, @anmolbabu , I get a 404 from docker when I use hub.docker.com/devanggaur7/supervisor as the image tag. It only works for me with docker.io . I can't test it locally otherwise.

File: Makefile
Comment (devang-gaur) [reply Makefile]: Should I revert back to "anmolbabu/supervisor" ? or to the new centos registry ?

File: src/api/routes.js
Comment (anmolbabu) [reply src/api/routes.js]: Ack.
Please make sure to log a issue for the route descriptor part we can triage it based on priority

File: package.json
Comment (anmolbabu) [reply package.json:22]: Ok

File: cico_setup.sh
Comment (anmolbabu) [reply cico_setup.sh]: Ok I didn't hit that issue and this worked for me.

Also, I only meant that after the script is verified and the PR is ready for merge, please retain back the old state of these lines as it would later be changed as in https://github.com/fabric8-hdd/supervisor/pull/38 once that's merged..

File: cico_setup.sh
Comment (anmolbabu) [reply cico_setup.sh]: I only meant that after the script is verified and the PR is ready for merge, please retain back the old state of these lines as it would later be changed as in #38 once that's merged..

File: Makefile
Comment (anmolbabu) [reply Makefile]: Revert back to old state the new changes will be done by #38

File: run_tests.sh
Comment (devang-gaur) [reply run_tests.sh:18]: Hey Anmol, I did explained the reason to @miteshvp .

The workflow is now like : 
1.) Build the image from Dockerfile.tests .
2.) If the build happens successfully, the test suite passed.
3.) The npm install in the Dockerfile.tests is inclusive of the testing and development utilitiesm which we don't need in the production environment.
4.) Now, since the image is not useful for us, as of now, we purge it to clear the memory space.

5.) We build the production ready fresh image in the Dockerfile and use npm install --production. Only the production required stuff goes in the container.

Apologies for the long comment. Let me know what do you think.

Regarding the stress testing, we can raise a seperate issue and PR when the time comes.

File: Makefile
Comment (anmolbabu) [inline Makefile:12]: This is intentional for the purpose of https://github.com/fabric8-hdd/supervisor/blob/master/cico_setup.sh#L21
anyway you could instead add make docker-build below before/after https://github.com/fabric8-hdd/supervisor/blob/master/cico_setup.sh#L22 as well

File: run_tests.sh
Comment (anmolbabu) [reply run_tests.sh:18]: Ok what I am trying to say is, if your intention is to have npm test run via docker image creation, we can as well do it without image creation right?
So this is only meaningful if we are doing some tests against the image or the image instantiated(running container)..
However, we'll take it in its current form and track the tests addition part as part of https://github.com/fabric8-hdd/supervisor/issues/42 ..
Please take care of my other comments and lets get this PR in ASAP

File: Makefile
Comment (devang-gaur) [reply Makefile:12]: Sure.

File: Dockerfile.tests
Comment (anmolbabu) [inline Dockerfile.tests:40]: Required?

File: Makefile
Comment (anmolbabu) [inline Makefile:19]: This docker build so you can probably have:
docker build -t supervisor-tests -f Dockerfile.tests .

File: cico_run_tests.sh
Comment (anmolbabu) [inline cico_run_tests.sh]: Can this be along the lines of https://github.com/fabric8-hdd/supervisor/pull/40/files#diff-b67911656ef5d18c4ae36cb6741b7965L23 and https://github.com/fabric8-hdd/supervisor/blob/master/cico_build_deploy.sh#L9

Comment (anmolbabu) [reply cico_run_tests.sh]: so that the single point of update for this kind of things would then be the makefile

File: Dockerfile.tests
Comment (devang-gaur) [reply Dockerfile.tests:40]: Hey, this would be required when I add stress testing scripts.

File: Makefile
Comment (devang-gaur) [reply Makefile:19]: No. The CACHEBUST argument is a necessary part here and I couldn't find any other way with Makefile to perform this.

You cannot assign any value to the variables in a Makefile code on run time. Its the only reason I resorted for this way. I've spent a good time to look for other ways already.

File: Dockerfile.tests
Comment (anmolbabu) [reply Dockerfile.tests:40]: Remove it for now add it back in the PR that requires this

**Comment by devang-gaur** [comment]
Changes made.

File: cico_run_tests.sh
Comment (anmolbabu) [inline cico_run_tests.sh:7]: @echo $(REGISTRY)/$(REPOSITORY)-tests:$(DEFAULT_TAG) add this in makefile and use that here. Otherwise in its current state, doesn’t it name the test image also as the main image?

**Review by anmolbabu — CHANGES_REQUESTED** [review]

File: cico_run_tests.sh
Comment (devang-gaur) [reply cico_run_tests.sh:7]: The names are same but tags are different. Simpler for us. We don't neet to maintain different versions of the test image. We just need one image. So, tagging as $IMAGE_NAME:test made sense to me.

Comment (anmolbabu) [reply cico_run_tests.sh:7]: No it’s not simple the image in the local cache will be overwritten

**Comment by devang-gaur** [comment]
Hey, @anmolbabu , changes made.

**Review by anmolbabu — APPROVED** [review]

**Pull request merged**
